File: server/lib/visits.js
const fs = require("fs");
const path = require("path");
const crypto = require("crypto");

const { DATA_DIR } = require("./paths");
const VISITS_FILE = path.join(DATA_DIR, "visits.json");
const MAX_VISITS = 500;

function ensureFile() {
  if (!fs.existsSync(DATA_DIR)) fs.mkdirSync(DATA_DIR, { recursive: true });
  if (!fs.existsSync(VISITS_FILE)) {
    fs.writeFileSync(VISITS_FILE, JSON.stringify({ visits: [] }, null, 2));
  }
}

function readVisits() {
  ensureFile();
  try {
    const data = JSON.parse(fs.readFileSync(VISITS_FILE, "utf8"));
    return Array.isArray(data.visits) ? data.visits : [];
  } catch {
    return [];
  }
}

function writeVisits(visits) {
  ensureFile();
  fs.writeFileSync(VISITS_FILE, JSON.stringify({ visits }, null, 2));
}

function parseUserAgent(ua = "") {
  const s = ua.toLowerCase();
  let os = "Неизвестно";
  if (s.includes("windows")) os = "Windows";
  else if (s.includes("iphone") || s.includes("ipad")) os = "iOS";
  else if (s.includes("android")) os = "Android";
  else if (s.includes("mac os") || s.includes("macintosh")) os = "macOS";
  else if (s.includes("linux")) os = "Linux";

  let browser = "Неизвестно";
  if (s.includes("edg/")) browser = "Edge";
  else if (s.includes("chrome/") && !s.includes("chromium")) browser = "Chrome";
  else if (s.includes("firefox/")) browser = "Firefox";
  else if (s.includes("safari/") && !s.includes("chrome")) browser = "Safari";
  else if (s.includes("opera") || s.includes("opr/")) browser = "Opera";

  let deviceType = "Компьютер";
  if (s.includes("mobile") || s.includes("iphone") || (s.includes("android") && !s.includes("tablet"))) {
    deviceType = "Телефон";
  } else if (s.includes("ipad") || s.includes("tablet")) {
    deviceType = "Планшет";
  }

  return { os, browser, deviceType };
}

function getClientIp(req) {
  const forwarded = req.headers["x-forwarded-for"];
  if (typeof forwarded === "string") return forwarded.split(",")[0].trim();
  return req.socket?.remoteAddress || req.ip || "—";
}

function recordVisit(req, body = {}) {
  const ua = req.headers["user-agent"] || body.userAgent || "";
  const parsed = parseUserAgent(ua);

  let deviceType = body.deviceType || parsed.deviceType;
  if (body.screenWidth && body.screenWidth < 768) deviceType = "Телефон";
  else if (body.screenWidth && body.screenWidth < 1024 && deviceType === "Компьютер") {
    deviceType = "Планшет";
  }

  const visit = {
    id: crypto.randomBytes(8).toString("hex"),
    at: new Date().toISOString(),
    ip: getClientIp(req),
    path: body.path || "/",
    sessionId: body.sessionId || null,
    language: body.language || req.headers["accept-language"]?.split(",")[0] || "—",
    referrer: body.referrer || req.headers.referer || "—",
    screen: {
      width: body.screenWidth || null,
      height: body.screenHeight || null,
    },
    device: {
      type: deviceType,
      os: parsed.os,
      browser: parsed.browser,
    },
    userAgent: ua.slice(0, 300),
  };

  const visits = readVisits();
  visits.unshift(visit);
  writeVisits(visits.slice(0, MAX_VISITS));
  return visit;
}

function getVisitHistory(limit = 100) {
  return readVisits().slice(0, limit);
}

function clearVisits() {
  writeVisits([]);
}

function getVisitStats() {
  const visits = readVisits();
  const uniqueSessions = new Set(visits.map((v) => v.sessionId).filter(Boolean)).size;
  const byDevice = {};
  for (const v of visits) {
    const key = `${v.device.type} · ${v.device.os}`;
    byDevice[key] = (byDevice[key] || 0) + 1;
  }
  return {
    total: visits.length,
    uniqueSessions,
    byDevice: Object.entries(byDevice)
      .map(([name, count]) => ({ name, count }))
      .sort((a, b) => b.count - a.count),
  };
}

module.exports = {
  recordVisit,
  getVisitHistory,
  getVisitStats,
  clearVisits,
};

File: backend/app/services/visits.py
import secrets
from collections import Counter

from sqlalchemy.orm import Session

from ..models import Visit


def record_visit(db: Session, req, body: dict) -> dict:
    ua = body.get("userAgent") or req.headers.get("user-agent", "")
    s = ua.lower()
    device_os = "Неизвестно"
    if "windows" in s:
        device_os = "Windows"
    elif "iphone" in s or "ipad" in s:
        device_os = "iOS"
    elif "android" in s:
        device_os = "Android"
    elif "mac" in s:
        device_os = "macOS"
    elif "linux" in s:
        device_os = "Linux"

    device_browser = "Неизвестно"
    if "edg/" in s:
        device_browser = "Edge"
    elif "chrome/" in s and "chromium" not in s:
        device_browser = "Chrome"
    elif "firefox/" in s:
        device_browser = "Firefox"
    elif "safari/" in s and "chrome" not in s:
        device_browser = "Safari"

    device_type = body.get("deviceType") or "Компьютер"
    sw = body.get("screenWidth")
    if sw and sw < 768:
        device_type = "Телефон"
    elif sw and sw < 1024 and device_type == "Компьютер":
        device_type = "Планшет"

    forwarded = req.headers.get("x-forwarded-for", "")
    ip = forwarded.split(",")[0].strip() if forwarded else (req.client.host if req.client else "—")

    visit_id = secrets.token_hex(8)
    visit = Visit(
        id=visit_id,
        ip=ip,
        path=body.get("path", "/"),
        session_id=body.get("sessionId"),
        language=body.get("language") or req.headers.get("accept-language", "—").split(",")[0],
        referrer=body.get("referrer") or req.headers.get("referer", "—"),
        screen_width=body.get("screenWidth"),
        screen_height=body.get("screenHeight"),
        device_type=device_type,
        device_os=device_os,
        device_browser=device_browser,
        user_agent=ua[:300],
    )
    db.add(visit)
    db.commit()
    return {"id": visit_id}


def list_visits(db: Session, limit: int = 100) -> list[dict]:
    rows = db.query(Visit).order_by(Visit.created_at.desc()).limit(limit).all()
    return [
        {
            "id": v.id,
            "at": v.created_at.isoformat() + "Z",
            "ip": v.ip,
            "path": v.path,
            "sessionId": v.session_id,
            "language": v.language,
            "referrer": v.referrer,
            "screen": {"width": v.screen_width, "height": v.screen_height},
            "device": {
                "type": v.device_type,
                "os": v.device_os,
                "browser": v.device_browser,
            },
            "userAgent": v.user_agent,
        }
        for v in rows
    ]


def visit_stats(db: Session) -> dict:
    visits = db.query(Visit).all()
    sessions = {v.session_id for v in visits if v.session_id}
    counter = Counter(f"{v.device_type} · {v.device_os}" for v in visits)
    by_device = [{"name": k, "count": c} for k, c in counter.most_common()]
    return {
        "total": len(visits),
        "uniqueSessions": len(sessions),
        "byDevice": by_device,
    }


def clear_visits(db: Session) -> None:
    db.query(Visit).delete()
    db.commit()
